### PR TITLE
feat: 1155 extensions

### PR DESCRIPTION
**Pull request by joaquim-verges**
CNCT-570

---

## PR-Codex overview
This PR adds ERC1155 extensions for `claimTo`, `lazyMint`, and `setClaimConditions`. It also includes event definitions and utility functions for conversions and hashing.

### Detailed summary
- Added ERC1155 extensions: `claimTo`, `lazyMint`, `setClaimConditions`
- Event definitions for `TokensLazyMinted` and `TokensClaimed`
- Utility functions for conversions and hashing
- Updated type definitions and type imports

> The following files were skipped due to too many changes: `packages/thirdweb/package.json`, `packages/thirdweb/src/extensions/erc1155/write/lazyMint.ts`, `packages/thirdweb/src/merkletree/MerkleTree.test.ts`, `packages/thirdweb/src/utils/extensions/drops/process-override-list.ts`, `packages/thirdweb/src/utils/extensions/drops/fetch-proofs-for-claimers.ts`, `packages/thirdweb/src/utils/extensions/drops/get-multicall-set-claim-claim-conditon-transactions.ts`, `packages/thirdweb/src/extensions/erc1155/__generated__/ILazyMint/write/lazyMint.ts`, `packages/thirdweb/src/merkletree/MerkleTree.ts`, `packages/thirdweb/src/extensions/erc1155/drops/write/claimTo.ts`, `packages/thirdweb/src/extensions/erc1155/drop1155.test.ts`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

File: packages/thirdweb/biome.json [hunk 1]
```diff
@@ -6,7 +6,9 @@
       "src/crypto/aes/lib/md5.ts",
       "src/utils/promise/p-limit.ts",
       "src/utils/bytecode/cbor-decode.ts",
-      "src/**/__generated__/**"
+      "src/**/__generated__/**",
+      "src/merkletree/MerkleTree.ts",
```
Comment (jnsdls) [inline packages/thirdweb/biome.json:10]: i see u gave up

Comment (joaquim-verges) [reply packages/thirdweb/biome.json:10]: mainly a bunch of 're-assigned params', could maybe just drop that rule instead

File: packages/thirdweb/src/exports/react.ts [hunk 2]
```diff
@@ -19,6 +19,7 @@ export type {
 } from "../react/web/ui/ConnectWallet/ConnectWalletProps.js";
 export type { NetworkSelectorProps } from "../react/web/ui/ConnectWallet/NetworkSelector.js";
 export type { WelcomeScreen } from "../react/web/ui/ConnectWallet/screens/types.js";
+export type { LocaleId } from "../react/web/ui/types.js";
```
Comment (jnsdls) [inline packages/thirdweb/src/exports/react.ts:22]: bit of a weird addition here? fine but not part of the rest of the changes

File: packages/thirdweb/src/extensions/erc1155/drops/write/claimTo.ts [hunk 3]
```diff
@@ -95,45 +31,45 @@ export type ClaimToParams = {
...
-      // TODO implement fetching merkle data
+      let proofData = {
+        currency: cc.currency,
+        proof: [] as `0x${string}`[],
+        quantityLimitPerWallet: cc.quantityLimitPerWallet,
+        pricePerToken: cc.pricePerToken,
+      };
       if (cc.merkleRoot !== padHex("0x", { size: 32 })) {
-        throw new Error("Allowlisted claims not implemented yet");
+        const claimerProof = await fetchProofsForClaimer({
```
Comment (jnsdls) [inline packages/thirdweb/src/extensions/erc1155/drops/write/claimTo.ts:48]: should lazy load this only if we have a merkleRoot etc

File: packages/thirdweb/src/extensions/erc1155/drops/write/claimTo.ts [hunk 4]
```diff
@@ -95,45 +31,45 @@ export type ClaimToParams = {
...
+          contract: options.contract,
+          claimer: options.from || options.to, // receiver and claimer can be different, always prioritize the claimer for allowlists
+          merkleRoot: cc.merkleRoot,
+        });
+
+        if (claimerProof) {
+          proofData = claimerProof;
+        }
```
Comment (jnsdls) [inline packages/thirdweb/src/extensions/erc1155/drops/write/claimTo.ts:56]: this has code smell, I think we should just branch via return statements instead, the `let` is gonna cause problems here later

File: packages/thirdweb/src/extensions/erc1155/drops/write/setClaimConditions.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
+import type { BaseTransactionOptions } from "../../../../transaction/types.js";
+import { getMulticallSetClaimConditionTransactions } from "../../../../utils/extensions/drops/get-multicall-set-claim-claim-conditon-transactions.js";
```
Comment (jnsdls) [inline packages/thirdweb/src/extensions/erc1155/drops/write/setClaimConditions.ts:2]: I caused this file name but I hate it regardless

File: packages/thirdweb/src/extensions/erc721/write/lazyMint.ts [hunk 6]
```diff
@@ -49,9 +49,7 @@ export type LazyMintParams = {
  * });
  * ```
  */
-export async function lazyMint(
-  options: BaseTransactionOptions<LazyMintParams>,
-) {
+export function lazyMint(options: BaseTransactionOptions<LazyMintParams>) {
```
Comment (jnsdls) [inline packages/thirdweb/src/extensions/erc721/write/lazyMint.ts:52]: good fix!

File: packages/thirdweb/src/merkletree/Base.ts [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+   * print
+   * @desc Prints out a visual representation of the merkle tree.
+   * @example
+   *```js
+   *tree.print()
+   *```
+   */
+  print(): void {
+    Base.print(this);
+  }
```
Comment (jnsdls) [inline packages/thirdweb/src/merkletree/Base.ts:18]: do we truly need this?

File: packages/thirdweb/src/merkletree/Base.ts [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+   * @example
+   *```js
+   *MerkleTree.print(tree)
+   *```
+   */
+
+  // biome-ignore lint/suspicious/noExplicitAny: debug print
+  static print(tree: any): void {
+    console.log(tree.toString());
+  }
```
Comment (jnsdls) [inline packages/thirdweb/src/merkletree/Base.ts:237]: do we even need this?

File: packages/thirdweb/src/utils/extensions/drops/fetch-proofs-for-claimers.ts [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+  );
+  const tree = new MerkleTree(hashedEntries, keccak256, {
+    sort: true,
+  });
+  // 5. get the proof for the claimer + the sub merkle tree root
+  const entry = shardData.entries.find(
+    (i) => i.address.toLowerCase() === claimer.toLowerCase(),
+  );
+  if (!entry) {
+    return undefined;
```
Comment (jnsdls) [inline packages/thirdweb/src/utils/extensions/drops/fetch-proofs-for-claimers.ts:75]: todo @jnsdls consistent return type `null` or `undefined` pick one ser

**Review by jnsdls — CHANGES_REQUESTED** [review]
some changes

**Review by jnsdls — APPROVED** [review]

File: packages/thirdweb/src/exports/react.ts (hunk 2 above)
Comment (joaquim-verges) [reply packages/thirdweb/src/exports/react.ts:22]: was requested by a user

**Pull request merged**